server/src/instant/db/transaction.clj

```diff
@@ -141,36 +146,119 @@
      app-id
      [{:message (format "You can't make updates to this app.")}])))
 
-(defn enforce-on-deletes
-  "Unoptimized on-delete helper that will delete entities that have
-   on-delete=:cascade. Only used for system attributes right now.
-   Before releasing as a feature, this should probably happen inside of
-   the `delete-multi`.
-
-   This approach is not ideal because there's no bound on the number of
-   times you recur. We probably want to prevent cycles on the on-delete
-   (e.g. if `a` has on-delete for `b`, then `b` can't have on-delete
-   for `a`), so that the delete query can be statically defined in
-   triple-model/delete-multi!"
-  ([conn attrs app-id results]
-   (enforce-on-deletes conn attrs app-id results (:delete-entity results)))
-  ([conn attrs app-id results last-delete-entities]
-   (if-let [deleted-triples (seq last-delete-entities)]
-     (let [deletes (keep (fn [{:keys [triples/attr_id triples/entity_id]}]
-                           (let [attr (attr-model/seek-by-id attr_id attrs)]
-                             (when (and (= :ref (:value-type attr))
-                                        (= :cascade (:on-delete attr)))
-                               [entity_id (attr-model/fwd-etype attr)])))
-                         deleted-triples)]
-       (if (seq deletes)
-         (let [on-delete-results (triple-model/delete-entity-multi! conn app-id deletes)]
-           (recur conn
-                  attrs
-                  app-id
-                  (update results :delete-entity into on-delete-results)
-                  on-delete-results))
-         results))
-     results)))
+(defn resolve-lookups
+  "Given [[attr-id value] [attr-id value] ...],
+   returns {[attr-id value] eid,
+            [attr-id value] eid,
+            ...}"
+  [conn app-id lookups]
+  (if (empty? lookups)
+    {}
+    (let [lookups (map (fn [[a v]] [a (->json v)]) lookups)
+          query   {:with [[[:lookups {:columns [:attr_id :value]}]
+                           {:values lookups}]]
+                   :from :triples
+                   :where [:= :app_id app-id]
```

Comment:
You probably want to check for `:triples.av` either here or in the join to make the query faster (we know that a lookup has to be on a unique attr, so it gets the `av` index).